paddle/fluid/framework/parallel_executor.cc

```diff
@@ -0,0 +1,845 @@
+  std::string DebugString() const override {
+    std::stringstream ss;
+    ss << name_ << ":" << place_;
+    return ss.str();
+  }
+
+  // version field currently is not used, however, just store the version to
+  // debug easily.
+  size_t version_;
+  std::string name_;
+  platform::Place place_;
+};
+
+struct DummyVarHandle : public VarHandleBase {
+  std::string DebugString() const override { return "dummy"; }
+};
+
+struct DependencyVarHandle : public VarHandleBase {
+  std::string DebugString() const override { return "Dependency Variable"; }
+};
+
+struct OpHandle {
+  std::vector<VarHandleBase *> inputs_;
+  std::vector<VarHandleBase *> outputs_;
+  std::unordered_map<platform::Place, platform::DeviceContext *,
+                     platform::PlaceHash>
+      dev_ctx_;
+
+  std::string DebugString() {
+    std::stringstream ss;
+    ss << "(";
+    for (auto *var : inputs_) {
+      ss << var->DebugString() << ", ";
+    }
+    ss << ") --> (";
+    for (auto *var : outputs_) {
+      ss << var->DebugString() << ", ";
+    }
+    ss << ")\n";
+    return ss.str();
+  }
+
+  virtual ~OpHandle() {}
+
+  virtual void Run() { PADDLE_THROW("Not implemented"); }
+  virtual void Wait(platform::DeviceContext *waited_dev) {}
+};
+
+struct ComputationOpHandle : public OpHandle {
+  std::unique_ptr<OperatorBase> op_;
+  Scope *scope_;
+  platform::Place place_;
+  cudaEvent_t event_;
+
+  explicit ComputationOpHandle(const OpDesc &op_desc, Scope *scope,
+                               platform::Place place)
+      : op_(framework::OpRegistry::CreateOp(op_desc)),
+        scope_(scope),
+        place_(place) {
+    if (platform::is_gpu_place(place)) {
+      cudaSetDevice(boost::get<platform::CUDAPlace>(place_).device);
+      cudaEventCreateWithFlags(&event_, cudaEventDisableTiming);
+    }
+  }
+
+  ~ComputationOpHandle() {
+    // FIXME: Destroy Event
+  }
+
+  void Run() override {
+    // Wait other op if necessary
+    if (platform::is_gpu_place(place_)) {
+      int dev_id = boost::get<platform::CUDAPlace>(place_).device;
+      cudaSetDevice(dev_id);
+    }
+    auto *cur_ctx = dev_ctx_[place_];
+    for (auto *in : inputs_) {
+      if (in->generated_op_ && in->generated_op_->dev_ctx_[place_] != cur_ctx) {
+        in->generated_op_->Wait(cur_ctx);
+      }
+    }
+
+    op_->Run(*scope_, place_);
+    if (platform::is_gpu_place(place_)) {
+      auto stream = static_cast<platform::CUDADeviceContext *>(dev_ctx_[place_])
+                        ->stream();
+      PADDLE_ENFORCE(cudaEventRecord(event_, stream));
+    }
+  }
+
+  void Wait(platform::DeviceContext *waited_dev) override {
+    if (platform::is_cpu_place(waited_dev->GetPlace()) ||
+        platform::is_cpu_place(place_)) {
+      this->dev_ctx_.at(place_)->Wait();
+    } else {
+      auto stream =
+          static_cast<platform::CUDADeviceContext *>(waited_dev)->stream();
+      PADDLE_ENFORCE(cudaStreamWaitEvent(stream, event_, 0));
```

Comment:
Before an operator runs, it should wait all input variables' operators are done. 
So, here, we should wait for input variables's operators' event, not its own event.

Reply:
If the input variables are not done, it should be waited in `Run` method.

Reply:
```
void Run() override {
    // Wait other op if necessary
    if (platform::is_gpu_place(place_)) {
      int dev_id = boost::get<platform::CUDAPlace>(place_).device;
      cudaSetDevice(dev_id);
    }
    auto *cur_ctx = dev_ctx_[place_];
    for (auto *in : inputs_) {
      if (in->generated_op_ && in->generated_op_->dev_ctx_[place_] != cur_ctx) {
        in->generated_op_->Wait(cur_ctx, in->generated_op_->event_);
      }
    }

    op_->Run(*scope_, place_);
    if (platform::is_gpu_place(place_)) {
      auto stream = static_cast<platform::CUDADeviceContext *>(dev_ctx_[place_])
                        ->stream();
      PADDLE_ENFORCE(cudaEventRecord(event_, stream));
    }
  }

  void Wait(platform::DeviceContext *waited_dev,  cudaEvent_t event) override {
    if (platform::is_cpu_place(waited_dev->GetPlace()) ||
        platform::is_cpu_place(place_)) {
      this->dev_ctx_.at(place_)->Wait();
    } else {
      auto stream =
          static_cast<platform::CUDADeviceContext *>(waited_dev)->stream();
      PADDLE_ENFORCE(cudaStreamWaitEvent(stream, event, 0));
    }
  }
```

Reply:
I see... 
The codes in PR is good.